contracts/strategies/qv-nft-tiered/QVNftTieredStrategy.sol

```diff
@@ -0,0 +1,176 @@
+        );
+    }
+
+    /// @dev Internal initialize function that sets the poolId in the base strategy
+    function __QV_NFT_TieredStrategy_init(
+        uint256 _poolId,
+        bool _registryGating,
+        bool _metadataRequired,
+        ERC721[] memory _nfts,
+        uint256[] memory _maxVoiceCreditsPerNft,
+        uint256 _registrationStartTime,
+        uint256 _registrationEndTime,
+        uint256 _allocationStartTime,
+        uint256 _allocationEndTime
+    ) internal {
+        __QVSimpleStrategy_init(
+            _poolId,
+            _registryGating,
+            _metadataRequired,
+            0,
+            _registrationStartTime,
+            _registrationEndTime,
+            _allocationStartTime,
+            _allocationEndTime
+        );
+
+        uint256 nftsLength = _nfts.length;
+        if (nftsLength != _maxVoiceCreditsPerNft.length) {
+            revert INVALID();
+        }
+
+        for (uint256 i = 0; i < nftsLength;) {
+            ERC721 nft = _nfts[i];
+            nfts[i] = nft;
+            maxVoiceCreditsPerNft[nft] = _maxVoiceCreditsPerNft[i];
+            unchecked {
+                i++;
+            }
+        }
+    }
+
+    /// =========================
+    /// ==== View Functions =====
+    /// =========================
+
+    /// @notice Checks if the allocator is valid
+    /// @param _allocator The allocator address
+    /// @return true if the allocator is valid
+    function isValidAllocator(address _allocator) external view override returns (bool) {
+        return _isValidAllocator(_allocator);
+    }
+
+    /// =============================
+    /// ==== Internal Functions =====
+    /// =============================
+
+    function _isValidAllocator(address _allocator) internal view returns (bool) {
+        uint256 nftsLength = nfts.length;
+        for (uint256 i = 0; i < nftsLength;) {
+            if (nfts[i].balanceOf(_allocator) > 0) {
+                return true;
+            }
+
+            unchecked {
+                i++;
+            }
+        }
+
+        return false;
+    }
+
+    /// @notice Allocate votes to a recipient
+    /// @param _data The data
+    /// @param _sender The sender of the transaction
+    /// @dev Only the pool manager(s) can call this function
+    function _allocate(bytes memory _data, address _sender) internal override {
+        (address recipientId, ERC721 nft, uint256 nftId, uint256 voiceCreditsToAllocate) =
+            abi.decode(_data, (address, ERC721, uint256, uint256));
+
+        // check the voiceCreditsToAllocate is > 0
+        if (voiceCreditsToAllocate <= 0) {
+            revert INVALID();
+        }
+
+        // check the time periods for allocation
+        if (block.timestamp < allocationStartTime || block.timestamp > allocationEndTime) {
+            revert ALLOCATION_NOT_ACTIVE();
+        }
+
+        // check that the sender can allocate votes
+        if (!_isValidAllocator(_sender)) {
```

Comment:
can be replaced. we have to check if the sender owns this specific nft instead
if the nft was already used, line 156 will revert

Reply:
Thoughts?

```javascript
// NFT => NFT ID => Redeemed
mapping(ERC721 => mapping(uint256 => bool)) public redeemed;

// Setter - within _allocate
redeemed[nft][nftId] = true;

// Check
function _isValidAllocator(address _allocator, uint256 _nftId) internal view returns (bool) {
  uint256 nftsLength = nfts.length;
  for (uint256 i = 0; i < nftsLength;) {
      if (nfts[i].balanceOf(_allocator) > 0) {
          if (redeemed[nfts[i]][_nftId] == true) {
              return false;
          }
  
          return true;
      }
  
      unchecked {
          i++;
      }
  }
  
  return false;
  }
```

Reply:
@codenamejason 
this will not work for several reasons:

- If you have 5 NFTs and **only** the first from the list is redeemed, it will always be false, even if there are still voice credits available
- the function does not check if `_allocator` is owner of the `_nftId`. Therefore it is still possible to pass any `nftId`. If the`_allocator` owns nftId 420, he could also vote with nftIds 1, 2, 3, etc..

Reply:
i would suggest to simply replace `if (!_isValidAllocator(_sender))` with `if(nft.ownerOf(nftId) != sender)`